kernel/aster-nix/src/util/net/addr/family.rs

```diff
@@ -0,0 +1,270 @@
+    AF_NETBEUI = 13,
+    /// Security callback pseudo AF
+    AF_SECURITY = 14,
+    /// PF_KEY key management API
+    AF_KEY = 15,
+    AF_NETLINK = 16,
+    // Alias to emulate 4.4BSD
+    // AF_ROUTE = AF_NETLINK
+    /// Packet family
+    AF_PACKET = 17,
+    /// Ash
+    AF_ASH = 18,
+    /// Acorn Econet
+    AF_ECONET = 19,
+    /// ATM SVCs
+    AF_ATMSVC = 20,
+    /// RDS sockets
+    AF_RDS = 21,
+    /// Linux SNA Project (nutters!)
+    AF_SNA = 22,
+    /// IRDA sockets
+    AF_IRDA = 23,
+    /// PPPoX sockets
+    AF_PPPOX = 24,
+    /// Wanpipe API Sockets
+    AF_WANPIPE = 25,
+    /// Linux LLC
+    AF_LLC = 26,
+    /// Native InfiniBand address
+    AF_IB = 27,
+    /// MPLS
+    AF_MPLS = 28,
+    /// Controller Area Network
+    AF_CAN = 29,
+    /// TIPC sockets
+    AF_TIPC = 30,
+    /// Bluetooth sockets
+    AF_BLUETOOTH = 31,
+    /// IUCV sockets
+    AF_IUCV = 32,
+    /// RxRPC sockets
+    AF_RXRPC = 33,
+    /// mISDN sockets
+    AF_ISDN = 34,
+    /// Phonet sockets
+    AF_PHONET = 35,
+    /// IEEE802154 sockets
+    AF_IEEE802154 = 36,
+    /// CAIF sockets
+    AF_CAIF = 37,
+    /// Algorithm sockets
+    AF_ALG = 38,
+    /// NFC sockets
+    AF_NFC = 39,
+    /// vSockets
+    AF_VSOCK = 40,
+    /// Kernel Connection Multiplexor
+    AF_KCM = 41,
+    /// Qualcomm IPC Router
+    AF_QIPCRTR = 42,
+    /// smc sockets: reserve number for
+    /// PF_SMC protocol family that
+    /// reuses AF_INET address family
+    AF_SMC = 43,
+    /// XDP sockets
+    AF_XDP = 44,
+    /// Management component transport protocol
+    AF_MCTP = 45,
+}
+
+const ADDR_MAX_LEN: usize = 128;
+
+/// Storage that can contain _any_ socket addresses.
+///
+/// The size and layout of this structure is specified by RFC 3493. For details, see
+/// <https://datatracker.ietf.org/doc/html/rfc3493#section-3.10>.
+#[repr(C)]
+#[derive(Debug, Clone, Copy, Pod)]
+struct Storage {
+    sa_family: u16,
+    bytes: [u8; ADDR_MAX_LEN - 2],
+    _align: [u64; 0],
+}
+
+/// Reads a socket address from userspace.
+///
+/// This method returns `Err(EINVAL)` for invalid socket address lengths and `Err(EAFNOSUPPORT)`
+/// for unsupported address families.
+///
+/// These error codes may be different from the Linux ones, but the difference is tricky and hard
+/// to fix. The main reason for this is that in Linux it's up to each protocol to decide how to
+/// intercept the bytes that represent socket addresses, but here this method is designed to parse
+/// socket addresses before diving deep into protocol-specific code.
+pub fn read_socket_addr_from_user(addr: Vaddr, addr_len: usize) -> Result<SocketAddr> {
+    if addr_len > ADDR_MAX_LEN {
+        return_errno_with_message!(Errno::EINVAL, "the socket address length is too large");
+    }
+
+    if addr_len < 2 {
+        return_errno_with_message!(Errno::EINVAL, "the socket address length is too small");
+    }
+
+    let mut storage = Storage::new_uninit();
```

Comment:
I'm not sure about the future of `Pod::new_uninit`. It _can_ be implemented soundly using inline assembly, or using `MayUninit::freeze` (after the corresponding RFC lands). So I don't know if we plan to remove it or not.

I still write `new_uninit` here because if we decide to remove it, we can always easily and quickly replace all `new_uninit` with `new_zeroed` without causing any problems. (We can even use `new_zeroed` to implement `new_uninit` as a temporary solution and postpone the decision to the future).

If we're sure the method will be removed, I can also make the change in this PR.

Reply:
Yes, I'm sure this will be removed. 

The `pod` crate will be replaced by `zerocopy` as suggested by #1107.

`Pod::new_zeroed` have its counterpart in `zerocopy` named [FromZeroes](https://docs.rs/zerocopy/0.7.35/zerocopy/trait.FromZeroes.html), but `new_uninit` does not. So just using `new_zeroed` then it can be replaced with API from zerocopy.